File: LogiOS/TrustarApp/Database/result/Entity/CommonCollectionResult.swift
//
//  CommonCollectionResult.swift
//  TrustarApp
//
//  Created by CuongNguyen on 15/03/2022.
//  Copyright © 2022 DionSoftware. All rights reserved.
//

import Foundation
import GRDB

enum CommonCollectionResultColumns: String, ColumnExpression {
    case companyCd, userId, allocationNo, allocationRowNo, postJson
}

class CommonCollectionResult: BaseEntity, Codable {
    var companyCd: String = Resources.strEmpty
    var userId: String = Resources.strEmpty
    var allocationNo: String = Resources.strEmpty
    var allocationRowNo: Int = Resources.zeroNumber
    var postJson: [CommonRow] = []
    
    override init() {
        super.init()
    }

    override class var databaseTableName: String { "CommonCollectionResult" }
    
    override class var primaryKey: String { "companyCd,userId,allocationNo,allocationRowNo" }

    required init(row: Row) {
        companyCd = row[CommonCollectionResultColumns.companyCd]
        userId = row[CommonCollectionResultColumns.userId]
        allocationNo = row[CommonCollectionResultColumns.allocationNo]
        allocationRowNo = row[CommonCollectionResultColumns.allocationRowNo]
        postJson = Helper.Shared.convertJsonToObject(json: row[CommonCollectionResultColumns.postJson])
        super.init(row: row)
    }
    
    init(userInfo: UserInfo, result: CollectionResult) {
        companyCd = userInfo.companyCd
        userId = userInfo.userId
        allocationNo = result.allocationNo
        allocationRowNo = result.allocationRowNo
        
        let r = result.rows()
        for it in r {
            postJson.append(CommonRow(allocationNo: result.allocationNo,
                                      allocationRowNo: result.allocationRowNo,
                                      collectionNo: it.collectionNo,
                                      collectionCd: it.collectionCd,
                                      collectionNm: it.collectionNm,
                                      actualQuantity: it.actualQuantity,
                                      collectionClass: it.collectionClass))
        }
        
        super.init()
        self.recordChanged()
    }

    override func encode(to container: inout PersistenceContainer) {
        container[CommonCollectionResultColumns.companyCd] = companyCd
        container[CommonCollectionResultColumns.userId] = userId
        container[CommonCollectionResultColumns.allocationNo] = allocationNo
        container[CommonCollectionResultColumns.allocationRowNo] = allocationRowNo
        container[CommonCollectionResultColumns.postJson] = Helper.Shared.convertObjectToJson(body: postJson)
        super.encode(to: &container)
    }
    
    class CommonRow: Codable {
        var allocationNo: String
        var allocationRowNo: Int
        var collectionNo: String?
        var collectionCd: String?
        var collectionNm: String?
        var collectionCountRes: Double
        var collectionClass: Int
        
        init(allocationNo: String,
             allocationRowNo: Int,
             collectionNo: String?,
             collectionCd: String?,
             collectionNm: String?,
             actualQuantity: Double,
             collectionClass: Int
        ){
            self.allocationNo = allocationNo
            self.allocationRowNo = allocationRowNo
            self.collectionNo = collectionNo
            self.collectionCd = collectionCd
            self.collectionNm = collectionNm
            self.collectionCountRes = actualQuantity
            self.collectionClass = collectionClass
        }
    }
}

File: LogAndroid/TrustarApp/src/main/kotlin/jp/co/toukei/log/trustar/db/result/entity/CommonCollectionResult.kt
package jp.co.toukei.log.trustar.db.result.entity

import androidx.room.ColumnInfo
import androidx.room.Entity
import jp.co.toukei.log.lib.jsonArrNotNull
import jp.co.toukei.log.trustar.db.user.entity.CollectionResult
import jp.co.toukei.log.trustar.user.UserInfo
import org.json.JSONArray
import third.jsonObj

@Entity(
        tableName = "common_collection_result",
        primaryKeys = ["company_cd", "user_id", "allocation_no", "allocation_row_no"]
)
class CommonCollectionResult(
        companyCd: String,
        userId: String,
        @ColumnInfo(name = "allocation_no") @JvmField val allocationNo: String,
        @ColumnInfo(name = "allocation_row_no") @JvmField val allocationRowNo: Int,
        @ColumnInfo(name = "post_json") @JvmField val postJson: JSONArray, // No pk for rows.
) : CommonUserSync(companyCd, userId) {

    constructor(userInfo: UserInfo, result: CollectionResult) : this(
            userInfo.companyCd,
            userInfo.userId,
            result.allocationNo,
            result.allocationRowNo,
            result.rows().jsonArrNotNull {
                jsonObj {
                    "allocationNo" v result.allocationNo
                    "allocationRowNo" v result.allocationRowNo
                    "collectionNo" v it.collectionNo
                    "collectionCd" v it.collectionCd
                    "collectionNm" v it.collectionNm
                    "collectionCountRes" v it.actualQuantity
                    "collectionClass" v it.collectionClass
                }
            }
    )
}
